File: heuristic.py
from generateMove import GenerateMove
from table import table
import heuristicGenerateMove

class Heuristic(GenerateMove):
    def __init__(self, l, r, threads, range, width = 10000):
        super().__init__(threads)
        self.l = l
        self.r = r
        self.width = width
        self.range = range

    # similiar to dynamic approach but only considers localy best width results

    def generate_move(self, read):

        a = (len(read.signal)//(len(read.sequence)-5))
        self.l, self.r = max(1, a-self.range), a+2+2*self.range
        read.move = heuristicGenerateMove.generate_move(
            signal= read.signal,
            sequence= read.sequence,
            l = self.l,
            r = self.r,
            width = self.width
        )
        return read.move
        n = len(read.signal)
        k = len(read.sequence) - 5

        memo = [dict() for _ in range(k+1)]
        memo[0][0] = (0, 0, 0, -1)

        for t in range(k):
            can = list(memo[t].values())
            can.sort()
            for h, s, i, _ in can[:min(len(can), self.width)]:
                cur = 0
                kmer = read.sequence[t:t+6]
                for x in range(self.l):
                    if i+x >= len(read.signal): break
                    cur += (read.signal[i+x] - table[kmer]) ** 2
                for x in range(self.l, self.r):
                    if i+x > len(read.signal): break
                    if i+x + (k-t-1) * self.l <= n and i+x + (k-t-1) * (self.r-1) >= n:
                        if not i+x in memo[t+1]:
                            memo[t+1][i+x] = (123456789, 123456789, -1, -1)
                        memo[t+1][i+x] = min(memo[t+1][i+x], ((cur + s)/(i+x), cur+s, i+x, i))
                    if i+x >= len(read.signal): break
                    cur += (read.signal[i+x] - table[kmer]) ** 2
        read.move = [0 for _ in range(len(read.signal))]
        c = n
        # print(memo)
        try:
            for i in range(k, 1, -1):
                # print(c, i)
                # print('memo', memo[i][c])
                read.move[memo[i][c][3]] = 1
                c = memo[i][c][3]
        except:
            print('fail ', end='')
            print(memo)
            exit(1)

        print(read.move)
        print(heuristicGenerateMove.generate_move(
            signal= read.signal,
            sequence= read.sequence,
            l = self.l,
            r = self.r,
            width = self.width
        ))

        cppmove = heuristicGenerateMove.generate_move(
            signal= read.signal,
            sequence= read.sequence,
            l = self.l,
            r = self.r,
            width = self.width
        )


        assert(read.move == cppmove)

    def __str__(self):
        return f"heuristic r={self.range}, width={self.width}"

File: heuristic.cpp
#include <bits/stdc++.h>
#include "table.cpp"
using namespace std;

vector<double> seq_signal(string s){
    map<char, int> alp = {{'A', 0}, {'C', 1}, {'G', 2}, {'T', 3}};

    vector<double> res;
    int cur = 0;
    for(int i = 0; i < s.size(); i++)
    {
        cur *= 4;
        cur += alp[s[i]];
        if(i > 4)
        {
            res.push_back(table[cur]);
            cur %= 1024;
        }
    }
    return res;
}

struct DPstate{
    double avg_err;
    double sum_err;
    int index;
    int previous_index;

    DPstate(double a, double s, int i, int p){
        avg_err = a;
        sum_err = s;
        index = i;
        previous_index = p;
    }
};

bool operator<(const DPstate &a, const DPstate &b)
{
    return a.avg_err < b.avg_err;
}

vector<int> generate_move(vector<double> signal, string sequence, int l, int r, int width) {
    auto seq = seq_signal(sequence);
    int n = signal.size(), k = seq.size();
    vector<vector<DPstate>> memo(2, vector<DPstate>(signal.size()+1, DPstate(INFINITY, INFINITY, -1, -1)));
    memo[0][0] = DPstate(0, 0, 0, -1);
    map<pair<int, int>, int> old;


    for(int t = 0; t < k; t++)
    {
        int id = t&1;
        sort(memo[id].begin(), memo[id].end());
        memo[!id].assign(signal.size(), DPstate(INFINITY, INFINITY, -1, -1));
        int poc = width-1;
        for(auto state : memo[id])
        {
            poc--;
            if(!poc) break;
            if(state.index == -1) break;
            // cout << state.avg_err << ' ' << state.sum_err << ' ' << state.index << ' ' << state.previous_index << endl;
            int i = state.index;
            old[{t, i}] = state.previous_index;
            double cur_err = state.sum_err;
            for(int x = 0; x < l; x++)
            {
                if(i + x > n) break;
                cur_err += (signal[i+x] - seq[t])*(signal[i+x]-seq[t]);
            }
            for(int x = l; x < r; x++)
            {
                if(i+x > n) break;
                if(i+x + (k-t-1) * l <= n && i + x + (k-t-1) * (r-1) >= n)
                {
                    memo[!id][i+x] = min(memo[!id][i+x], DPstate(cur_err/(i+x), cur_err, i+x, i));
                    // cout << "penis\n" << i << ' ' << x << ' ' << n << endl;
                    // cout << cur_err << ' ' << memo[!id][i+x].avg_err << endl;

                }
                if(i+x >= n) break;
                cur_err += (signal[i+x] - seq[t])*(signal[i+x]-seq[t]);
                // // for(auto x : memo[!id]) cout << x.avg_err << ' ' << x.sum_err << ' ' << x.index << ' ' << x.previous_index << endl;
            }
        }
    }
    old[{k, n}] = memo[k&1][n].previous_index;

    // cout << "old\n";
    // for (auto const& [key, val] : old) cout << key.first << ' ' << key.second << ' ' << val << endl;
    
    // cout << '\n';
    // cout << 'k' << k << " n" << n << endl;

    vector<int> move(n, 0);
    int c = n;
    for(int i = k; i > 1; i--)
    {
        // cout << c << i << endl;
        move[old[{i, c}]] = 1;
        c = old[{i, c}];
    }

    return move;



    vector<int> res(signal.size(), 0);
    for(int i = 1; i <= sequence.size()-6; i++)
        res[i] = 1;
    return res;
}

